cmake: Set top-level target output locations

hebasto opened the pull request:
This PR sets the target output locations to the `bin` and `lib` subdirectories within the build tree, creating a directory structure that mirrors that of the installed targets.

This approach is widely adopted by the large projects, such as [LLVM](https://github.com/llvm/llvm-project/blob/e146c1867e8decfd423034f63a3a863733e03f04/lldb/cmake/modules/LLDBStandalone.cmake#L128-L130):
```cmake
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib${LLVM_LIBDIR_SUFFIX})
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib${LLVM_LIBDIR_SUFFIX})
```

The `libsecp256k1` project has also recently [adopted](https://github.com/bitcoin-core/secp256k1/pull/1553) this approach.

With this PR, all binaries are conveniently located. For example, run:
```
$ ./build/bin/fuzz
```
instead of:
```
$ ./build/src/test/fuzz/fuzz
```

On Windows, all required DLLs are now located in the same directory as the executables, allowing to run `bitcoin-chainstate.exe` (which loads `bitcoinkernel.dll`) without the need to copy DLLs or modify the `PATH` variable.

The idea was briefly discussed among the build team during the recent CoreDev meeting.

---

**Warning**: This PR changes build locations of newly built executables like `bitcoind` and `test_bitcoin` from `src/` to `bin/` without deleting previously built executables. A clean build is recommended to avoid accidentally running old binaries.

TheCharlatan commented:
Concept ACK

fanquake commented:
This seems fine, but it'd be good to primarily list the benefits to our project, rather than justifying making changes based on if some other open source project happens to do it, as other projects may have different constraints, or various reasoning for making the same change. Also, in this PR you say `This approach is widely adopted by the large projects, such as LLVM`, but in the secp256k1 thread for the same change you said that our current behaviour is ["the default behaviour adopted by many projects."](https://github.com/bitcoin-core/secp256k1/pull/1553#issuecomment-2360229365), so it'd seem like either way of doing things is mostly fine/normal.

My understanding is that this change is mostly a convenience change for native Windows developers, so that it becomes easier to run binaries after compilation (without installing?). Possibly with the potential for us to follow up with / remove rpath usage entirely (cc @theuni). What I don't really understand is how every other project that uses CMake as we do now, has solved this issue, if they haven't made the same change as in this PR (or why this wouldn't be the default CMake behaviour, if it otherwise results in broken (native) Windows binaries).

fanquake commented:
Also added this to `29.x`, as if we are going to change this, it needs to happen along with the CMake switchover, and not be a new breaking change after that.

theuni commented:
Concept ACK.

As @fanquake mentioned above, we currently have:
```cmake
  set_target_properties(bitcoin-chainstate PROPERTIES
    SKIP_BUILD_RPATH OFF
  )
```

Which I think can be removed now, no?

hebasto commented:
> Concept ACK.
> 
> ... we currently have:
> 
> ```cmake
>   set_target_properties(bitcoin-chainstate PROPERTIES
>     SKIP_BUILD_RPATH OFF
>   )
> ```
> 
> Which I think can be removed now, no?

No, we can't. Only for Windows builds, `*.exe` and `*.dll` files reside in a single `bin` subdirectory. For other platforms, shared libraries still reside in their own `lib` subdirectory, which still require using RPATH.

No, we can't. Only for Windows builds `*.exe` and `*.dll` files reside in a single `bin` subdirectory. On other platforms, shared libraries still reside in their own `lib` subdirectory, which requires the continued use of RPATH.

> Edit: Also, I think this obsoletes the `CMAKE_SKIP_BUILD_RPATH` comment in `CMakeLists.txt`.

I believe this comment is still relevant.

hebasto commented:
Rebased due to the conflict with the merged bitcoin/bitcoin#31191.

BrandonOdiwuor reviewed (commented):
Concept ACK

theuni commented:
> > Which I think can be removed now, no?
> 
> No, we can't. Only for Windows builds `*.exe` and `*.dll` files reside in a single `bin` subdirectory. On other platforms, shared libraries still reside in their own `lib` subdirectory, which requires the continued use of RPATH.
> 
> > Edit: Also, I think this obsoletes the `CMAKE_SKIP_BUILD_RPATH` comment in `CMakeLists.txt`.
> 
> I believe this comment is still relevant.

Yes, you're right on both counts. I forgot about the different lib path for Windows (which is what makes it work there). Thanks for clarifying.

hebasto commented:
Rebased due to the conflict with the merged https://github.com/bitcoin/bitcoin/pull/26593.

theStack reviewed (commented):
Did some quick testing on Ubuntu 22.04 LTS and OpenBSD 7.6, verifying that the binaries end up in a single place as expected:
```
$ ls ./build/bin/                                                                                                             
bitcoin-cli      bitcoin-util     bitcoind         noverify_tests   test_bitcoin     unitester
bitcoin-tx       bitcoin-wallet   exhaustive_tests object           tests
```
Not sure how people feel about binaries from subtree projects also being placed there, especially if they have very generic names like "tests" or "object"? (I thought first that https://github.com/bitcoin-core/secp256k1/pull/1582 would also add the `secp256k1_` prefix to the binaries, but apparently it only affects test names for `ctest`...)

hebasto commented:
Rebased on top of the https://github.com/bitcoin/bitcoin/pull/31307 to workaround MSVC compiler issue.

ryanofsky reviewed (approved):
Code review ACK 06ff9429d32943aa4debec38ea105c8af4283aec. It seems like a improvement to me overall if binaries in the build directory have the same layout as binaries in releases. I'm not sure current implementation of this is ideal but it seems like a a step in the right direction.

Having binaries built in consistent locations that match install locations should make it is easier to find them, make scripts and wrappers simpler, and avoid bugs that only happen in developer builds but not final installs or vice-versa.

Things I think would be good to improve:

- It would be good if binaries from subtrees were handled differently from our own binaries. See
[#31161 (review)](https://github.com/bitcoin/bitcoin/pull/31161#pullrequestreview-2443913213). I think ideally subtrees would be unaffected by this change and their binaries would remain in same locations they would be if subtrees were built separately.
- It would be good if multiprocess binaries `bitcoin-node` and `bitcoin-gui` could be built in libexec/ instead of than bin/ since with [#30975](https://github.com/bitcoin/bitcoin/pull/30975) these binaries are meant to be installed to libexec/ and not present in the PATH or called directly by users.
- Possibly, it might make sense to build bitcoin-chainstate.exe and bitcoinkernel.dll to libexec/ instead of bin/.
- Maybe it would be good to temporarily add symlinks from old locations to new locations to avoid breaking developers workflows abruptly and make it easier to migrate. Currently this PR seems to be a breaking change: [#31161 (comment)](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2444839927)

re: https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2444644193

> What I don't really understand is how every other project that uses CMake as we do now, has solved this issue, if they haven't made the same change as in this PR (or why this wouldn't be the default CMake behaviour, if it otherwise results in broken (native) Windows binaries)

I think most CMake projects avoid the library-finding issue by having a flatter hierarchy, and not using `add_subdirectory` calls as much. This avoids tying binary locations to source layout, and puts binaries in one place instead of scattering them in subdirectories.

I think the reason for bin/ and lib/ build directories not being cmake default behavior is also out of a preference for flatness and not having unnecessary nesting, but also because cmake is meant to be platform agnostic and tries to follow conventions of each platform. It wouldn't use bin/ and lib/ directories on windows because that is a unix convention.

The reasons for cmake default behavior don't really apply in our case though, since we are already building things in subdirectories, and our windows releases already follow unix conventions by having a top level archive directory and bin/ and share/ subdirectories.

hebasto commented:
@ryanofsky 

> Code review ACK [06ff942](https://github.com/bitcoin/bitcoin/commit/06ff9429d32943aa4debec38ea105c8af4283aec). It seems like a improvement to me overall if binaries in the build directory have the same layout as binaries in releases. I'm not sure current implementation of this is ideal but it seems like a a step in the right direction.
> 
> Having binaries built in consistent locations that match install locations should make it is easier to find them, make scripts and wrappers simpler, and avoid bugs that only happen in developer builds but not final installs or vice-versa.

Thank you for your thoughtful review!

> Things I think would be good to improve:
> 
>    * It would be good if binaries from subtrees were handled differently from our own binaries. See
>       [#31161 (review)](https://github.com/bitcoin/bitcoin/pull/31161#pullrequestreview-2443913213). I think ideally subtrees would be unaffected by this change and their binaries would remain in same locations they would be if subtrees were built separately.

Reworked. Additionally, I treat the `univalue` ex-subtree the same way as other subtrees because of its test binaries. I doubt if we are interested in locating them in `bin/`.

>    * It would be good if multiprocess binaries `bitcoin-node` and `bitcoin-gui` could be built in libexec/ instead of than bin/ since with [#30975](https://github.com/bitcoin/bitcoin/pull/30975) these binaries are meant to be installed to libexec/ and not present in the PATH or called directly by users.

Thanks! implemented.

>    * Possibly, it might make sense to build bitcoin-chainstate.exe and bitcoinkernel.dll to libexec/ instead of bin/.

I don't have a strong opinion about that. Leaving it as is for now.

>    * Maybe it would be good to temporarily add symlinks from old locations to new locations to avoid breaking developers workflows abruptly and make it easier to migrate. Currently this PR seems to be a breaking change: [#31161 (comment)](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2444839927)

TBH, I don't like the idea of temporary code. If this PR lands before the v29 branch-off, the potential disruption will be limited.

hebasto commented:
A silent conflict has been fixed.

fanquake commented:
https://cirrus-ci.com/task/6223844022681600?logs=ci#L2898:
```bash
[08:40:37.725] + LD_LIBRARY_PATH=/ci_container_base/depends/i686-pc-linux-gnu/lib
[08:40:37.725] + test/functional/test_runner.py --ci -j10 --tmpdirprefix /ci_container_base/ci/scratch/test_runner/ --ansi --combinedlogslen=99999999 --timeout-factor=40 --v2transport --quiet --failfast
[08:40:38.317] 2025-01-10T13:40:38.202000Z TestFramework (INFO): PRNG seed is: 5569585375096173377
[08:40:38.317] 2025-01-10T13:40:38.206000Z TestFramework (INFO): Initializing test directory /ci_container_base/ci/scratch/test_runner/test_runner_₿_🏃_20250110_134037/cache
[08:40:38.317] 2025-01-10T13:40:38.211000Z TestFramework (ERROR): Unexpected exception caught during testing
[08:40:38.317] Traceback (most recent call last):
[08:40:38.317]   File "/ci_container_base/test/functional/test_framework/test_framework.py", line 131, in main
[08:40:38.317]     self.setup()
[08:40:38.317]   File "/ci_container_base/test/functional/test_framework/test_framework.py", line 314, in setup
[08:40:38.317]     self.setup_chain()
[08:40:38.317]   File "/ci_container_base/test/functional/test_framework/test_framework.py", line 405, in setup_chain
[08:40:38.317]     self._initialize_chain()
[08:40:38.317]   File "/ci_container_base/test/functional/test_framework/test_framework.py", line 875, in _initialize_chain
[08:40:38.317]     self.start_node(CACHE_NODE_ID)
[08:40:38.317]   File "/ci_container_base/test/functional/test_framework/test_framework.py", line 575, in start_node
[08:40:38.317]     node.start(*args, **kwargs)
[08:40:38.317]   File "/ci_container_base/test/functional/test_framework/test_node.py", line 256, in start
[08:40:38.317]     self.process = subprocess.Popen(self.args + extra_args, env=subp_env, stdout=stdout, stderr=stderr, cwd=cwd, **kwargs)
[08:40:38.317]                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[08:40:38.317]   File "/usr/lib/python3.12/subprocess.py", line 1026, in __init__
[08:40:38.317]     self._execute_child(args, executable, preexec_fn, close_fds,
[08:40:38.317]   File "/usr/lib/python3.12/subprocess.py", line 1955, in _execute_child
[08:40:38.317]     raise child_exception_type(errno_num, err_msg, err_filename)
[08:40:38.317] FileNotFoundError: [Errno 2] No such file or directory: 'bitcoin-node'
[08:40:38.317] 2025-01-10T13:40:38.271000Z TestFramework (INFO): Stopping nodes
```

Sjors commented:
Concept ACK

Lightly tested 8c44a947b501c8417454527637bd3adb3ce2ff4e by building on macOS 15.2 (M4). I like how the documentation is changed with a scripted-diff. The build system changes in the first commit look reasonable to me at first glance, but did not review thoroughly.

@ryanofsky wrote:

> Possibly, it might make sense to build bitcoin-chainstate.exe and bitcoinkernel.dll to libexec/ instead of bin/.

I like that as well. If we add them to the release later, they won't clutter `bin/`.

Now is also a good time to move `bitcoin-util` there, because it's only used by our test code (and for calibrating a new signet).

Re symlinks: I would also prefer to avoid that, if we can get this into v29.

theuni commented:
> * It would be good if multiprocess binaries `bitcoin-node` and `bitcoin-gui` could be built in libexec/ instead of than bin/

This makes sense to me if we're going to go down the route of the wrapper executable, but I don't think we should do that until users aren't actually supposed to be launching them directly.

IMO it's not a problem to move them after a release cycle or two if the expected user behavior changes, if nothing else that's actually a good indication that their behavior _should_ change.

Sjors commented:
Ok, so now we have `bitcoin-util` and `bitcoin-chainstate` in `libexec`, with `bitcoin-gui` and `bitcoin-node` back in `bin` - which can be moved by #31375. That seems fine to me.

When I do `sudo cmake --install build` on macOS 15.2 it puts the binaries in `/usr/local/bin` and `/usr/local/lib` (not `/usr/local/libexec` or `/usr/libexec`). Is that intended?

hebasto commented:
Reworked per recent [feedback](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2586587671).

`libexec` is an installation path for targets whose output path is set to `libexec`.

fanquake commented:
Why are `bitcoin-chainstate` and `bitcoin-util` being put into libexec? `bitcoin-util` is meant to be called by users directly. Not sure why `bitcoin-chainstate` would be put there either?

theuni commented:
> Why are `bitcoin-chainstate` and `bitcoin-util` being put into libexec? `bitcoin-util` is meant to be called by users directly. Not sure why `bitcoin-chainstate` would be put there either?

Agreed. The point of libexec is (implicitly) to make it HARDER to run binaries from there, because they don't live in the user's `$PATH`. They're meant to be found by other binaries that _do_ live in `$PATH`. So long as we're recommending/relying on users running a binary by hand, it shouldn't live in libexec.

So I propose dropping the `libexec` stuff entirely for this PR. I _do_ think it's worth moving the binaries there in the future when the execution model changes. Like I said above, IMO moving them later would actually be a good thing, as it will force users into the updated behavior.

hebasto commented:
@fanquake @theuni 

Thank you for your feedback!

> So I propose dropping the `libexec` stuff entirely for this PR.

Dropped.

ryanofsky reviewed (approved):
Code review ACK e9edf28ba4ff4ebf02ae90874794c51c55ef789e. Nice changes! Again I think using consistent build locations that match install locations should make it is easier to see what binaries are available, simplify scripts and wrappers, and avoid bugs that only happen in developer builds but not final installs or vice-versa.

Since last review the main change seems to be that subproject binaries are not affected anymore by this PR, which is great.

I don't understand the details of the different libexec pushes. It seems ok to postpone using libexec for now, but I disagree with @theuni "So long as we're recommending/relying on users running a binary by hand, it shouldn't live in libexec."

IMO, it only makes sense for high level binaries intended to be used by normal users to be installed in bin/ and placed on the system PATH:

- bitcoin-cli
- bitcoind
- bitcoin-qt
- bitcoin-tx
- bitcoin-util
- bitcoin-wallet

I don't think it makes sense for the following binaries which are not intended to by used by typical users to be installed in bin/ and put into the PATH:

- bench_bitcoin
- bitcoin-chainstate (*)
- bitcoin-gui
- bitcoin-node
- fuzz (*)
- test_bitcoin
- test_bitcoin-qt

It seems better to either not install these binaries or to put them out of the way in libexec/ to avoid confusion. (It looks like the ones marked with an asterisk above are not currently installed, but the rest are.)

Would like to see more discussion about this and find out where the disagreement is so binaries don't have to move twice, but this doesn't need to block the PR.

hebasto commented on CMakeLists.txt:
```diff
@@ -585,9 +585,6 @@ set(CMAKE_SKIP_INSTALL_RPATH TRUE)
 add_subdirectory(test)
 add_subdirectory(doc)
 
-include(cmake/crc32c.cmake)
-include(cmake/leveldb.cmake)
-include(cmake/minisketch.cmake)
```
> Unclear whether it's necessary for this PR, or maybe it's just nice to consolidate these together with univalue?

The latter.

hebasto commented on src/CMakeLists.txt:
```diff
@@ -8,28 +8,17 @@ include(AddWindowsResources)
+include(crc32c)
+include(leveldb)
+include(minisketch)
 add_subdirectory(univalue)
```
The `include`d subprojects are subtrees managed with our custom CMake scripts. While there was an initial plan to switch to their upstream CMake scripts, this approach is currently [considered questionable](https://github.com/bitcoin/bitcoin/pull/30905).

Conversely, `univalue` was promoted from a subtree to our own codebase, adopting the same per-directory approach.

hebasto commented:
> Code review ACK [e9edf28](https://github.com/bitcoin/bitcoin/commit/e9edf28ba4ff4ebf02ae90874794c51c55ef789e). Nice changes! Again I think using consistent build locations that match install locations should make it is easier to see what binaries are available, simplify scripts and wrappers, and avoid bugs that only happen in developer builds but not final installs or vice-versa.

@ryanofsky 

Thank you for your review! The branch has been updated per your feedback.

hebasto commented on src/CMakeLists.txt:
```diff
@@ -79,6 +68,41 @@ set_target_properties(secp256k1 PROPERTIES
+endif()
+
+configure_file(${PROJECT_SOURCE_DIR}/cmake/bitcoin-build-config.h.in bitcoin-build-config.h USE_SOURCE_PERMISSIONS @ONLY)
+include_directories(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR})
```
Thanks! [Fixed](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2596961825).

hebasto commented on src/CMakeLists.txt:
```diff
@@ -8,28 +8,17 @@ include(AddWindowsResources)
+# unmodified.
+include(crc32c)
+include(leveldb)
+include(minisketch)
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2596961825).

hebasto commented on src/CMakeLists.txt:
```diff
@@ -8,28 +8,17 @@ include(AddWindowsResources)
+# Subprojects include subdirectories that do or could have tests
+# and/or benchmark binaries, such as all subtrees and univalue.
+# The top-level target output locations for subprojects remain
+# unmodified.
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2596961825).

ryanofsky reviewed (approved):
Code review ACK 186680acef661bd139bf3d16f494365ea55993b5. Since last review just reverted CMAKE_MODULE_PATH change, removed duplicate code and changed comment as suggested. Looks good!

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/31161#pullrequestreview-2556956641

> It seems better to either not install these binaries or to put them out of the way in libexec/ to avoid confusion. [...] Would like to see more discussion about this and find out where the disagreement is so binaries don't have to move twice, but this doesn't need to block the PR.

I opened #31679 to move internal binaries to libexec, since using libexec isn't directly related to what this PR is trying to do. Would probably make sense for more discussion about libexec to happen in #31679 instead of here.

theuni reviewed (approved):
This is much simpler than the last time I looked. Thanks ryanofsky for all the feedback here. We can argue about libexec in the other PR :)

utACK 186680acef661bd139bf3d16f494365ea55993b5

theStack reviewed (approved):
Lightly tested ACK 186680acef661bd139bf3d16f494365ea55993b5

Nice! Tested on Ubuntu 22.04.3 LTS with a fresh build using the `dev-mode` preset and verifying the build/{bin,lib} folder contents. Also checked that the only remaining binaries and libraries in build/src are from subtree projects (which makes sense to not mix it). Only skimmed quickly over the actual build system changes, as I'm not familiar with the ins and outs of CMake.

achow101 commented:
~0

I've read through the conversation but I still don't understand why this change is beneficial?

This is also a major breaking change and would break a significant amount of external tooling that rely on the current location of the binaries.

theuni commented:
@achow101 This is one of the changes I was referring to in last week's IRC meeting. We'd rather not move these around after v29 (with the exception of some moves to libexec) so that nobody has to migrate to CMake and then adapt again.

Not sure what you mean about breaking existing tooling though? This affects the location in the CMake build tree. It's not related to installation locations. So afaik the only tooling that needs to be updated has been changed in this PR. Presumably anyone else relying on the old autotools locations will be adapting for CMake anyway, and imo this makes migration even easier as now everything's in one place.

ryanofsky commented:
> ~0
> 
> I've read through the conversation but I still don't understand why this change is beneficial?

The main thing I like about this change is that makes it easier to write scripts and navigate the build. Examples:

1. When you run `make WHATEVER`, `WHATEVER` will be built in `bin/` if it is an executable `lib/` if it is a binary. You don't have to go looking all over the tree for the `WHATEVER` you just built in `src/` or `src/bench/` or `src/test/` or `src/qt/` or `/src/qt/test` or whatever random subdirectory the `add_executable` or `add_library` call is placed in. Currently `add_executable` and `add_library` calls are not placed in consistent locations because some subdirectories have CMakeLists.txt files and other don't and are built by their parents, and there is usually some reasoning behing this but as a build system user the placement seems arbitrary and random.

2. This makes builds match releases, so for example if you want to write a script that calls binaries you don't have to go hunting around different locations and hardcoding multiple paths to make it work in a both a build directory and a release. You can just point the script at the bin/ directory, not have hardcoded build/src/ paths.

3. It makes the build easier to navigate as a new user or developer. You can extract the source, run make, run `ls -l bin/` and see all the tools that have been built and are available to use. You don't have to hunt through the build/src/ directory and sift through object and makefile cruft to see tools that are available.

4. Even though we aren't really using shared libraries I think this change would especially make sense in case we did provide more shared libraries in the future. It is useful to be able to relocate builds and run them with LD_LIBRARY_PATH, and if all libraries are built in the same `lib/` that is easy to do. On windows there are no rpaths, so shared libraries actually need to be placed in the same folder (bin/) to work together easily. They can't be scattered in different build directories and just work.

> This is also a major breaking change and would break a significant amount of external tooling that rely on the current location of the binaries.

I think it would break internal tooling not external tooling since this PR does not affect install or release locations. Earlier I suggested this could create symlinks in old locations for convenience and to prevent breakage. I don't think too many tools should actually break, but I do think this change could be annoying to developers, for example if you just update git and rebuild, and try to run the old binary because you don't know the new binary has a different location.

achow101 commented:
> Not sure what you mean about breaking existing tooling though?

> I think it would break internal tooling not external tooling since this PR does not affect install or release locations. Earlier I suggested this could create symlinks in old locations for convenience and to prevent breakage. I don't think too many tools should actually break, but I do think this change could be annoying to developers, for example if you just update git and rebuild, and try to run the old binary because you don't know the new binary has a different location.

Yes, the tooling for developers is what I meant.

I have several scripts which aid me in building and testing prs. These already have to figure out when to use autotools and when to use cmake. With this PR, these would now all need to figure out where the binaries live, and that seems to be significantly less trivial to do than detecting autotools or cmake.

While I'm not opposed to this change, I'm not thrilled by it either. It feels a lot like a "it'd be nice if we did that" kind of change, but one that also ends up breaking a bunch of developer workflows while doing it. I would've preferred that this was done with cmake, or immediately after it, than now where we're a few weeks before feature freeze for 29.0.

ryanofsky commented:
> > Not sure what you mean about breaking existing tooling though?
>
> I have several scripts which aid me in building and testing prs. These already have to figure out when to use autotools and when to use cmake. With this PR, these would now all need to figure out where the binaries live, and that seems to be significantly less trivial to do than detecting autotools or cmake.

I think it would be nice not just for tools but also for developers to just create symlinks from the old to new locations. Here is a commit which does that, mostly written by chatgpt (if that is not obvious): 433aa99d238740bbdcb07b53b17f639170bd326d

hebasto commented:
>With this PR, these would now all need to figure out where the binaries live, and that seems to be significantly less trivial to do than detecting autotools or cmake.

Something like this should work:
```sh
[ -e build/bin/test_bitcoin ] && ln -s $(realpath build/bin/test_bitcoin) build/src/test/
```

> I would've preferred that this was done with cmake, or immediately after it, than now...

I agree with you in hindsight.

> ... where we're a few weeks before feature freeze for 29.0.

The point is to avoid breaking changes for the (external) users.

> I think it would be nice not just for tools but also for developers to just create symlinks from the old to new locations. Here is a commit which does that, mostly written by chatgpt (if that is not obvious): [433aa99](https://github.com/bitcoin/bitcoin/commit/433aa99d238740bbdcb07b53b17f639170bd326d)

I'd refrain form adding the code, which will effectively become dead in a couple of weeks or so. Also, the suggested commit might not work on Windows, where symbolic links are not enabled by default.

ryanofsky commented:
I think if this PR is going to be merged without providing backward compatibility, something should be done to warn developers, like adding a tag `[BREAKING]` or `[INCOMPATIBLE]` or `[!]` or emoji ‼️ or ⚠️ to the subject and a clear warning in the description and maybe saying something in IRC channel too.

In its current form, this change seems likely to cause confusion and wasted time because when upgrading or downgrading around the change, **builds will appear to succeed but old binaries will be left in place** so it will be easy to accidentally wind up testing with binaries you think contain changes that aren't there.

IMO, there is isn't a good reason to break compatibility here at all, if a shim that just adds one new line and one new file to the build https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2623350369 can provide it.

As long as we're going to customize output locations, putting symlinks in the default locations pointing to our custom locations, seems like a sensible thing to do. This would be useful for developers upgrading in near future, as well as developers testing or bisecting later on, and for developers who may be familiar with cmake but not familiar with our particular build and looking for outputs in the default cmake locations.

(Sorry for repeating this suggestion again. Just wanted to be clear about reasons I think it would be worthwhile.)

hebasto commented:
> IMO, there is isn't a good reason to break compatibility here at all, if a shim that just adds one new line and one new file to the build [#31161 (comment)](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2623350369) can provide it.

Thanks! Applied.

theuni commented on cmake/module/GenerateSymlinks.cmake:
```diff
@@ -0,0 +1,88 @@
+        return()
+    endif()
+
+    add_custom_target(${target_name}_create_symlink
```
I wonder if this should be an `add_custom_command()` with a [`POST_BUILD` event](https://cmake.org/cmake/help/latest/command/add_custom_command.html#build-events) instead? Presumably this version (I haven't checked) creates the symlink even if the link fails, whereas I would expect a `POST_BUILD`command to not have that problem.

ryanofsky replied:
What problem do you see this solving? Creating a symlink pointing at a target that does not exist should be fine. It will still point to the right place.

theuni reviewed (approved):
utACK 433aa99d238740bbdcb07b53b17f639170bd326d

Only change since my last review is the symlink script, which I don't feel strongly about either way. I left a comment about a fixup for it, but it's not a blocker for me.

fanquake commented:
When would we remove these symlinks? This seems like overkill, and just deferring breakage until later? CMake is already a huge disruption, and it's not clear that we need to add even more workarounds to accomodate a few developers before we've even managed to actually fully make the changeover. Some infra/script/fuzzing etc will need to be updated, but this all seems maangeable.

> Also, the suggested commit might not work on Windows, where symbolic links are not enabled by default.

Did anyone test this on Windows?

theuni commented:
> When would we remove these symlinks? This seems like overkill, and just deferring breakage until later? CMake is already a huge disruption, and it's not clear that we need to add even more workarounds to accomodate a few developers before we've even managed to actually fully make the changeover. Some infra/script/fuzzing etc will need to be updated, but this all seems maangeable.

I agree with this. We've not really attempted any autotools compatibility, I'm not sure why this would be an exception.

I just didn't want to see this PR blocked forever.

ryanofsky commented:
> When would we remove these symlinks? This seems like overkill, and just deferring breakage until later? 

> We've not really attempted any autotools compatibility

I don't see any reason to delete the symlinks in the future. And they are not created for compatibility with autotools but for compatibility with cmake defaults and compatibility with our own build.

Also, if we did decide to delete the symlinks in the future it would not be deferring the breakage but replacing **silent breakage** (i.e. you rebuild with change and your changes have no effect because you are running an old binary in there is no indication about what could be wrong) with **obvious breakage** (i.e. the location you were expecting the binary to exist does not exist at all).

IMO, there's no reason to a inflict silent breakage like this on developers if you are maintaining a build system. It should be a point of pride if you work on a build system for your system to just work and not waste other developers time.

ryanofsky commented:
> Did anyone test this on Windows?

Of course would be nice to test this on windows but it should work there according to cmake release notes: https://cmake.org/cmake/help/v3.13/release/3.13.html?highlight=release%20notes#command-line

fanquake commented:
> Of course would be nice to test this on windows but it should work

@hebasto ?

hebasto commented:
> > Did anyone test this on Windows?
> 
> Of course would be nice to test this on windows but it should work there according to cmake release notes: https://cmake.org/cmake/help/v3.13/release/3.13.html?highlight=release%20notes#command-line

The success of the `cmake -E create_symlink` command on Windows depends on the the "Create symbolic links" privilege ([`SeCreateSymbolicLinkPrivilege`](https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-10/security/threat-protection/security-policy-settings/create-symbolic-links)), which is disabled by default. There are a few ways to [enabled](https://neacsu.net/posts/win_symlinks/) it. It is enabled in GitHub Actions images.

With `SeCreateSymbolicLinkPrivilege` disabled, the build fails:
```
  Creating symlink from C:/Users/hebasto/bitcoin/build/src to C:/Users/hebasto/bitcoin/build/bin
CUSTOMBUILD : CMake error : failed to create symbolic link 'C:/Users/hebasto/bitcoin/build/src/bitcoin-cli.exe': A required privilege is not held by the client. [C:\Users\hebasto\bitcoin\build\src\bitcoin-cli_create_symlink.vcxproj]
```

hebasto commented:
Additionally, 433aa99d238740bbdcb07b53b17f639170bd326d does not work for any multi-config generators.

For example:
```
$ cmake -B build -G "Ninja Multi-Config"
$ cmake --build build
$ file build/src/bitcoind
build/src/bitcoind: broken symbolic link to /home/hebasto/git/bitcoin/build/bin/bitcoind

hebasto commented:
> > IMO, there is isn't a good reason to break compatibility here at all, if a shim that just adds one new line and one new file to the build [#31161 (comment)](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2623350369) can provide it.
> 
> Thanks! Applied.

Reverted back due to the GenerateSymlinks module was not robust enough.


> In its current form, this change seems likely to cause confusion and wasted time because when upgrading or downgrading around the change, **builds will appear to succeed but old binaries will be left in place** so it will be easy to accidentally wind up testing with binaries you think contain changes that aren't there.

It can be easily mitigated by starting a build with a clean build tree.

TheCharlatan commented:
Thought a guix build might be prudent since we are changing a bunch of paths (built on aarch64):
```
b683fabbea0455a7f8a5c0566079b1bbce438edd5870e11209ce2dd5a016541e  guix-build-186680acef66/output/aarch64-linux-gnu/SHA256SUMS.part
4c144bdb301bf7d81dd99a8cef2082950fd5912f8edba6cea545482445cf7dba  guix-build-186680acef66/output/aarch64-linux-gnu/bitcoin-186680acef66-aarch64-linux-gnu-debug.tar.gz
c4e4a24a847dd420cd1b9310e13aef1e6376d069c5099f799918dd3f2c3a01e3  guix-build-186680acef66/output/aarch64-linux-gnu/bitcoin-186680acef66-aarch64-linux-gnu.tar.gz
19abaa002bc6111ec792c2a014f4f3702f46444ff133c53a1c8fae96dfbf515b  guix-build-186680acef66/output/arm-linux-gnueabihf/SHA256SUMS.part
39c526797543e8daa8cc27d5bd35d9583266e555bc0ee4c20de56b64816b9f5d  guix-build-186680acef66/output/arm-linux-gnueabihf/bitcoin-186680acef66-arm-linux-gnueabihf-debug.tar.gz
04fb06febddf4bd049659e513d7f6732a6bf37cbc63a00c651c65b0056e2d00b  guix-build-186680acef66/output/arm-linux-gnueabihf/bitcoin-186680acef66-arm-linux-gnueabihf.tar.gz
80b5f686bfc8dd18200bca54f1aa7c96f58922a63885b3631276436c7f9bf076  guix-build-186680acef66/output/arm64-apple-darwin/SHA256SUMS.part
19087881d8ec9c0c0ad3d0d4b4f6e411b6b1f4f3fd5b1d24bf5b20ee911e710d  guix-build-186680acef66/output/arm64-apple-darwin/bitcoin-186680acef66-arm64-apple-darwin-unsigned.tar.gz
1527ddabd3bd54d7ae487c48d294ef93f196321b1eab056beced2cba594b9e64  guix-build-186680acef66/output/arm64-apple-darwin/bitcoin-186680acef66-arm64-apple-darwin-unsigned.zip
722aec652e082b53d4883eaa3f8265f0dcc465eea3489cd7699062a29844b206  guix-build-186680acef66/output/arm64-apple-darwin/bitcoin-186680acef66-arm64-apple-darwin.tar.gz
9f45d6238ebad1458260532e08ffe7e94327bc7a1f62d085a0d059f7afec51a5  guix-build-186680acef66/output/dist-archive/bitcoin-186680acef66.tar.gz
ac289f117d402574cbb39348051e1511b5c16d828d30bb80cdd2a9602814280e  guix-build-186680acef66/output/powerpc64-linux-gnu/SHA256SUMS.part
720e5189cf195b7705704db40cdcc913a16850b4dcd395786c5d050e1264c739  guix-build-186680acef66/output/powerpc64-linux-gnu/bitcoin-186680acef66-powerpc64-linux-gnu-debug.tar.gz
ccf1bcff094cf570c7c0c1cdc9ed6437ecf6f5a1c53460121195fd5d5e54833a  guix-build-186680acef66/output/powerpc64-linux-gnu/bitcoin-186680acef66-powerpc64-linux-gnu.tar.gz
c52dbba75e79a0b60c8f126eb885c58544966f9e6208eed8119236d6fc0f8c6d  guix-build-186680acef66/output/riscv64-linux-gnu/SHA256SUMS.part
31574a143f3f9871a730a172a3a75e16f3dd1c30d45648e7a9bfe5ed07384810  guix-build-186680acef66/output/riscv64-linux-gnu/bitcoin-186680acef66-riscv64-linux-gnu-debug.tar.gz
f55284b9922439667ef5ffbd7f44f519b42342562507df04874bf833f127fd0a  guix-build-186680acef66/output/riscv64-linux-gnu/bitcoin-186680acef66-riscv64-linux-gnu.tar.gz
e86ddbde80f1476ee02d6c110fc804ba268bf3f93fc3670be3833891dc453e50  guix-build-186680acef66/output/x86_64-apple-darwin/SHA256SUMS.part
2980cf2969d547f4c98c0229be5154b0e713857a68bc1bbbff60b3ef93169923  guix-build-186680acef66/output/x86_64-apple-darwin/bitcoin-186680acef66-x86_64-apple-darwin-unsigned.tar.gz
1c4a4fe448ddf315417eb1b8fe779ce11c230215c44ff81985f7e2eadc1ae7df  guix-build-186680acef66/output/x86_64-apple-darwin/bitcoin-186680acef66-x86_64-apple-darwin-unsigned.zip
0cc419d4bdca0c2c4ff298967210bc4643f982dbea2e14fde9e74da6ff8a44de  guix-build-186680acef66/output/x86_64-apple-darwin/bitcoin-186680acef66-x86_64-apple-darwin.tar.gz
a2eea9c6f98a35b9201645ab71c7a404ba4128ea2632cb48ff622527e36fc36e  guix-build-186680acef66/output/x86_64-linux-gnu/SHA256SUMS.part
e46f1db4f8455e6402fa3abc0b0ad1cf696e47754843880940a10855a590abd2  guix-build-186680acef66/output/x86_64-linux-gnu/bitcoin-186680acef66-x86_64-linux-gnu-debug.tar.gz
70079f253749293abb2f4c0cc4e06eb45735b1ab590670ae56cab9b5e4daad96  guix-build-186680acef66/output/x86_64-linux-gnu/bitcoin-186680acef66-x86_64-linux-gnu.tar.gz
4f51cb67ce09625cee37f5d6d87ad9c2362cf4e1a89b296fedb208ea8f1655ec  guix-build-186680acef66/output/x86_64-w64-mingw32/SHA256SUMS.part
0e7b0e458e8dc392cfb8cc61b3fe6aceb95927af41af317fff455e7f2772983d  guix-build-186680acef66/output/x86_64-w64-mingw32/bitcoin-186680acef66-win64-debug.zip
f701b3291b1bf7cd5134a55bfda7478001d662814d14de4e9f0bda6ef92853cd  guix-build-186680acef66/output/x86_64-w64-mingw32/bitcoin-186680acef66-win64-setup-unsigned.exe
c285c2254c9609591ccd85c68758a7454652fdcd57a34acca8f9b67371334d84  guix-build-186680acef66/output/x86_64-w64-mingw32/bitcoin-186680acef66-win64-unsigned.tar.gz
f80a2e0e51146de565191371cb9411655e6007123cd1a385c13f619e16d265e6  guix-build-186680acef66/output/x86_64-w64-mingw32/bitcoin-186680acef66-win64.zip
```

hebasto commented:
Rebased due to the conflict with the merged bitcoin/bitcoin#31268.

hebasto committed:
cmake: Set top-level target output locations

This change:
1. Collects build artifacts in dedicated locations.
2. Allows running bitcoin-chainstate.exe with bitcoinkernel.dll directly
from the build tree on Windows.

hebasto committed:
scripted-diff: Adjust documentation per top-level target output location

-BEGIN VERIFY SCRIPT-

ren() { sed -i "s|\<$1\>|$2|g" $( git grep -l "$1" :\(exclude\)./src/secp256k1 ) ; }

ren build/src/bench   build/bin
ren build/src/test    build/bin
ren build/src/qt/test build/bin
ren build/src/qt      build/bin
ren build/src         build/bin
ren build_fuzz/src/test/fuzz build_fuzz/bin

-END VERIFY SCRIPT-

hebasto commented:
My Guix build:
```
aarch64
66e46fc7610a8e3a39b6253672b5063ea0f1c3ea54df6807c8e1d36d416e9d8c  guix-build-f7d8a44ce42c/output/aarch64-linux-gnu/SHA256SUMS.part
6333dc4e4233970e5ec19ac3e3b99cc30eaedf78a5ec74da57363e0c6d9993fc  guix-build-f7d8a44ce42c/output/aarch64-linux-gnu/bitcoin-f7d8a44ce42c-aarch64-linux-gnu-debug.tar.gz
4225ab5948b61dcc631346c08d1a6afd6263afaaecaf9301098f1edfd2933964  guix-build-f7d8a44ce42c/output/aarch64-linux-gnu/bitcoin-f7d8a44ce42c-aarch64-linux-gnu.tar.gz
d10bfa76580551723ab09b5480b076bfd86c5ef602c0b01ea5188984a00efbba  guix-build-f7d8a44ce42c/output/arm-linux-gnueabihf/SHA256SUMS.part
955f2a94b817a6b96e9d87e31ec79857e373ebb81e28bc6578622c4c74857464  guix-build-f7d8a44ce42c/output/arm-linux-gnueabihf/bitcoin-f7d8a44ce42c-arm-linux-gnueabihf-debug.tar.gz
26fc0cfa35e4ec146f26c0ba99381f67c6e338e47baa1e96cee87b81c4787402  guix-build-f7d8a44ce42c/output/arm-linux-gnueabihf/bitcoin-f7d8a44ce42c-arm-linux-gnueabihf.tar.gz
b75469c57748f0ab3b2d2b297e1bb1ea3cc27cee8d9093e5b87fba3776ba7128  guix-build-f7d8a44ce42c/output/arm64-apple-darwin/SHA256SUMS.part
7e9c9e8b1e1acec4d23eb1dba10002aa0bb429a446e32ecdf199508365dbfbb5  guix-build-f7d8a44ce42c/output/arm64-apple-darwin/bitcoin-f7d8a44ce42c-arm64-apple-darwin-unsigned.tar.gz
4aecaa93639e6a30903c2669b4019c2a06c652dadf6a566d9e6cfdf408dda638  guix-build-f7d8a44ce42c/output/arm64-apple-darwin/bitcoin-f7d8a44ce42c-arm64-apple-darwin-unsigned.zip
4240dfe30b5451d7418604257ed4d833a5da579054eedd781e9fca1383b75deb  guix-build-f7d8a44ce42c/output/arm64-apple-darwin/bitcoin-f7d8a44ce42c-arm64-apple-darwin.tar.gz
fb4f101c6b3dc4d51e8630c87967c4a97bc73ae5ebe9f897f7b269f51cb18d03  guix-build-f7d8a44ce42c/output/dist-archive/bitcoin-f7d8a44ce42c.tar.gz
137322a0267467cca1aed1cd8c47f70a82576b978ac5bf6ecf8a9aa542977043  guix-build-f7d8a44ce42c/output/powerpc64-linux-gnu/SHA256SUMS.part
0ad51a212e6ce3b9c6136ea9899c65805a6870c6f509d4e982816cdc2482ca20  guix-build-f7d8a44ce42c/output/powerpc64-linux-gnu/bitcoin-f7d8a44ce42c-powerpc64-linux-gnu-debug.tar.gz
511151ca1dfe5823e9a0294fc6c833ba03cf4b4eb5ef1cec3d8b583d3ecf19d6  guix-build-f7d8a44ce42c/output/powerpc64-linux-gnu/bitcoin-f7d8a44ce42c-powerpc64-linux-gnu.tar.gz
e590fb03009078692ac6ce5dbb9684ab4025d246936cd4d3aa5cf9b02401d92d  guix-build-f7d8a44ce42c/output/riscv64-linux-gnu/SHA256SUMS.part
e3017e204efdace62ccc54e62cf99e5484de584341cbc8fbbd03e4dab006897f  guix-build-f7d8a44ce42c/output/riscv64-linux-gnu/bitcoin-f7d8a44ce42c-riscv64-linux-gnu-debug.tar.gz
fc1a4a0e86b23701342e20e7cb947fc23568b9894bff213138fe746d8f4e2614  guix-build-f7d8a44ce42c/output/riscv64-linux-gnu/bitcoin-f7d8a44ce42c-riscv64-linux-gnu.tar.gz
edc9631762747b108c71f21eb771c1dcb0d2266c4b5d70f4007121bbb71576ce  guix-build-f7d8a44ce42c/output/x86_64-apple-darwin/SHA256SUMS.part
5441ba677a439aa83ca40d2e3bff1674c22519b4f73f4b4f6861847145c6476b  guix-build-f7d8a44ce42c/output/x86_64-apple-darwin/bitcoin-f7d8a44ce42c-x86_64-apple-darwin-unsigned.tar.gz
f7ff0b0c316cb3e57a38c8f32933054c9bd109dd7205ba2763721ed42de17103  guix-build-f7d8a44ce42c/output/x86_64-apple-darwin/bitcoin-f7d8a44ce42c-x86_64-apple-darwin-unsigned.zip
545ca13e7c71401733f427fd4a390ea9f118eb7bb8cf285d439babe26433aec6  guix-build-f7d8a44ce42c/output/x86_64-apple-darwin/bitcoin-f7d8a44ce42c-x86_64-apple-darwin.tar.gz
56a64b532acb1f09d1ec6beb7ab00615da409d6f930a363b79ec6681838b8ea5  guix-build-f7d8a44ce42c/output/x86_64-linux-gnu/SHA256SUMS.part
0baf6d98d4eefb85e9075588ee0e0900a1df0f7b2c916dbc88791c5fc5736991  guix-build-f7d8a44ce42c/output/x86_64-linux-gnu/bitcoin-f7d8a44ce42c-x86_64-linux-gnu-debug.tar.gz
2a967ac73582220d5b61047e919ad8e400cdc1799bffa51fd25c3094320f7d0a  guix-build-f7d8a44ce42c/output/x86_64-linux-gnu/bitcoin-f7d8a44ce42c-x86_64-linux-gnu.tar.gz
ebc76d2f3b4c872077132e96f6741bff8bba8274c47b98975fbc5bda823e8c33  guix-build-f7d8a44ce42c/output/x86_64-w64-mingw32/SHA256SUMS.part
3de4c2fa6f379f83ad0edc22ef7c626b5388a2a640d88339adc12a90b99eeaa6  guix-build-f7d8a44ce42c/output/x86_64-w64-mingw32/bitcoin-f7d8a44ce42c-win64-debug.zip
cc117cbb7f13bb6c2f25f36b49637a5f88d25d813f78a176ea7ee16bc922bb2d  guix-build-f7d8a44ce42c/output/x86_64-w64-mingw32/bitcoin-f7d8a44ce42c-win64-setup-unsigned.exe
ea191b9991a27fc7288560e82967f88006bbd88e2feb37de8696859ff3c6a9fb  guix-build-f7d8a44ce42c/output/x86_64-w64-mingw32/bitcoin-f7d8a44ce42c-win64-unsigned.tar.gz
ef98f5411f66d3910b6936c910c08dde0ea7485088b60576c8bc6246a00611f7  guix-build-f7d8a44ce42c/output/x86_64-w64-mingw32/bitcoin-f7d8a44ce42c-win64.zip
```

TheCharlatan commented on src/CMakeLists.txt:
```diff
@@ -7,28 +7,18 @@ include(AddWindowsResources)
-add_subdirectory(crypto)
+#=============================
+# Subprojects
+#=============================
```
Nit: These headers in the comments seem a bit inconsistent at the moment. What makes the secp subtree special enough to get its own header? Where do these sections end? Why don't the other components get their own headers?

hebasto replied:
> What makes the secp subtree special enough to get its own header?

For other subtrees, it has not yet been decided whether their own build scripts will be used.

fanquake replied:
> it has not yet been decided whether their own build scripts will be used.

The current decision is to not use them (i.e https://github.com/bitcoin/bitcoin/pull/30905). As far as I'm aware, no one is working on the changes that would be required for us to start using them, so I'm not sure why we'd start using them.

sedited reviewed (approved):
ACK f7d8a44ce42cce0b3010a7f7d14ec180e959f3e4

theStack reviewed (approved):
Lightly tested re-ACK f7d8a44ce42cce0b3010a7f7d14ec180e959f3e4

(retested as described in https://github.com/bitcoin/bitcoin/pull/31161#pullrequestreview-2578563294 before)

ryanofsky reviewed (approved):
Code review ACK 879569cab4e5b400350f3b95d7bee71b49636591 (no changes since last review other than rebase), but [again](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2631914588) I think this PR should be prefixed with [BREAKING] or [INCOMPATIBLE] or [!] or ‼️ or ⚠️ and have a clear message to developers about what it does like "**Warning**: This PR changes build locations of newly built executables like `bitcoind` and `test_bitcoin` from `src/` to `bin/` without deleting previously built executables. A clean build is recommended to avoid accidentally running old binaries."

re: https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2650448405

> Reverted, as the `GenerateSymlinks` module was not sufficiently robust.

I still think a change like 433aa99d238740bbdcb07b53b17f639170bd326d would be a strict improvement to avoid creating confusion and silent failures when updating. And to be compatible with cmake default build locations. If symlinks are disabled on windows bitcoin builds or in multi-config ninja bitcoin builds that seems harmless because I don't think we have any developers regularly using these.

> > In its current form, this change seems likely to cause confusion and wasted time because when upgrading or downgrading around the change, **builds will appear to succeed but old binaries will be left in place** so it will be easy to accidentally wind up testing with binaries you think contain changes that aren't there.
>
> It can be easily mitigated by starting a build with a clean build tree.

The point is that failures can be silent, so you don't know there is anything to mitigate. For example, you make a change that causes a segfault but you don't know it causes a segfault because you are running a stale binary. Or you are trying to fix a bug and your fixes do not work because you do not know you are a running a stale binary.

I don't think this breakage should block the PR if others are ok with it. I'm only arguing against it because it seems unnecessary.

hebasto commented:
> Code review ACK [879569c](https://github.com/bitcoin/bitcoin/commit/879569cab4e5b400350f3b95d7bee71b49636591) (no changes since last review other than rebase), but [again](https://github.com/bitcoin/bitcoin/pull/31161#issuecomment-2631914588) I think this PR should be prefixed with [BREAKING] or [INCOMPATIBLE] or [!] or ‼️ or ⚠️ and have a clear message to developers about what it does like "**Warning**: This PR changes build locations of newly built executables like `bitcoind` and `test_bitcoin` from `src/` to `bin/` without deleting previously built executables. A clean build is recommended to avoid accidentally running old binaries."

Done.

hebasto commented:
Rebased due to the conflict with the merged bitcoin/bitcoin#31742.

ryanofsky commented:
> I don't think it is a supported use-case to leave the cmake build dir stale and append builds on it, without fully deleting it first.

What are you talking about? I am doing this constantly many times every day, when I am checking out different PRs and branches and running make and expecting binaries to be up to date when it finishes. If other developers are happy to break this assumption I'm also happy though. Just trying to point out alternatives and tradeoffs here.

Sjors commented:
> I am doing this constantly many times every day, when I am checking out different PRs

Same, but I've sometimes found it unreliable, just as with autotools. For example on this branch, if you first do `-DENABLE_WALLET=OFF` and then `-DENABLE_WALLET=ON` if will fail with "Wallet functionality requested but no BDB or SQLite support available."

Baking a guix build...

No strong opinion on the symlinks.

theStack reviewed (approved):
Light re-ACK 568fcdddaec2cc8decba5a098257f31729cc1caa

hebasto commented:
My Guix build:
```
aarch64
d94fbbb8bed0a91349c7962d108e0296d6cfef0f4438f9918026fe3e284f5a5a  guix-build-568fcdddaec2/output/aarch64-linux-gnu/SHA256SUMS.part
a126b070d9baac09d151e1de17fa21e0f8283f43f0b8bce4c4d53af86f2fdc99  guix-build-568fcdddaec2/output/aarch64-linux-gnu/bitcoin-568fcdddaec2-aarch64-linux-gnu-debug.tar.gz
5a082deacad4c4d970315b6f03bbc9fe7aae2de46cc99709502838b4487d0c57  guix-build-568fcdddaec2/output/aarch64-linux-gnu/bitcoin-568fcdddaec2-aarch64-linux-gnu.tar.gz
696a670e4e3684c461c007f0959a62e1cce6cb21c804c44c74c9d8df20ba9095  guix-build-568fcdddaec2/output/arm-linux-gnueabihf/SHA256SUMS.part
65bd4b76cb84bd6c049d6bd82d16c2ad3d923500d783235ff666580f96889892  guix-build-568fcdddaec2/output/arm-linux-gnueabihf/bitcoin-568fcdddaec2-arm-linux-gnueabihf-debug.tar.gz
9d5189fee684c9e719d7059796d8cfb37afd3f64a79df610d07eca988633d9d7  guix-build-568fcdddaec2/output/arm-linux-gnueabihf/bitcoin-568fcdddaec2-arm-linux-gnueabihf.tar.gz
1b64e1395a95b09baf1de9f506b8a688f54af219fee336e1f44c8ed4be895a04  guix-build-568fcdddaec2/output/arm64-apple-darwin/SHA256SUMS.part
53f1d88f28c7b324335f6ca1fdbe5387a3a37363a5b4aa70239323b8de699a76  guix-build-568fcdddaec2/output/arm64-apple-darwin/bitcoin-568fcdddaec2-arm64-apple-darwin-unsigned.tar.gz
f0e3db976de94f7f4408d326f00769da8d5e4e48ca68427890b5e1cfda15fc69  guix-build-568fcdddaec2/output/arm64-apple-darwin/bitcoin-568fcdddaec2-arm64-apple-darwin-unsigned.zip
d175a54842bc12d1a71d595c6bfc7f7d9e98115acbb3238edacb4baba9e3988e  guix-build-568fcdddaec2/output/arm64-apple-darwin/bitcoin-568fcdddaec2-arm64-apple-darwin.tar.gz
d4b3a9cbb39981b530141912a4045ee03a547faeb90a9fe1a6fadeb1083b75fe  guix-build-568fcdddaec2/output/dist-archive/bitcoin-568fcdddaec2.tar.gz
5ed2c32ca39566129facc08584a33ea3508a00053a4a6934eb61446d33c577ee  guix-build-568fcdddaec2/output/powerpc64-linux-gnu/SHA256SUMS.part
bda8d954d121d6850f46ab4c9d654a10e8d5ecfc1d72f15a0ca47291e419626c  guix-build-568fcdddaec2/output/powerpc64-linux-gnu/bitcoin-568fcdddaec2-powerpc64-linux-gnu-debug.tar.gz
324a9d5ce128d8e42e78d5f6343564235f5e0ddfc7d32828658275a381d43457  guix-build-568fcdddaec2/output/powerpc64-linux-gnu/bitcoin-568fcdddaec2-powerpc64-linux-gnu.tar.gz
d062e44e0bc843dab5150490b67a96bb6ecf8e4fe9d610d9019caa656078d927  guix-build-568fcdddaec2/output/riscv64-linux-gnu/SHA256SUMS.part
db1225ef9bfa52a696435e02c6eefacd3225caeb5b0af5a275a211a923801f07  guix-build-568fcdddaec2/output/riscv64-linux-gnu/bitcoin-568fcdddaec2-riscv64-linux-gnu-debug.tar.gz
a08b5d0bdc0ec600713b72ceae9bbd6a0cd83c6639ad99e056311b24812bad8d  guix-build-568fcdddaec2/output/riscv64-linux-gnu/bitcoin-568fcdddaec2-riscv64-linux-gnu.tar.gz
0872cb0255d98002fb366e836d4cda21af4ca6fa7d218dc2a0bf7e6a38cbc205  guix-build-568fcdddaec2/output/x86_64-apple-darwin/SHA256SUMS.part
2a30e149b91f88deda7afb4031ddb490dc7103cc2984df924dd32a81eb5b4b29  guix-build-568fcdddaec2/output/x86_64-apple-darwin/bitcoin-568fcdddaec2-x86_64-apple-darwin-unsigned.tar.gz
e208a478504c3a70f209745cf490a7c509a995b0ff4e613b6b7b592073ab99fc  guix-build-568fcdddaec2/output/x86_64-apple-darwin/bitcoin-568fcdddaec2-x86_64-apple-darwin-unsigned.zip
c7339e8169bb55d81dc8d34f26b4ab9e924c3b0a811406525b11f5a8cdd82065  guix-build-568fcdddaec2/output/x86_64-apple-darwin/bitcoin-568fcdddaec2-x86_64-apple-darwin.tar.gz
0cd0eb8c35576ab23955249385b351dde5e452bf57eb21492457cc8df1145b23  guix-build-568fcdddaec2/output/x86_64-linux-gnu/SHA256SUMS.part
b5ce7824d75a06a23109c060a42666531cecc2cca714ac3f85469f830ec5b266  guix-build-568fcdddaec2/output/x86_64-linux-gnu/bitcoin-568fcdddaec2-x86_64-linux-gnu-debug.tar.gz
59e92b3ca351c00656a34222f97e2a4cc7bf8a479100f1322fc45140252df240  guix-build-568fcdddaec2/output/x86_64-linux-gnu/bitcoin-568fcdddaec2-x86_64-linux-gnu.tar.gz
ae5f9141796c4d8926b73d4157549f298ca1dc6f913d77e916a2d126311bcb13  guix-build-568fcdddaec2/output/x86_64-w64-mingw32/SHA256SUMS.part
d979830bcff6f684102b5d5e783d6357166f3b416b4d33cc3f716a48203a40a8  guix-build-568fcdddaec2/output/x86_64-w64-mingw32/bitcoin-568fcdddaec2-win64-debug.zip
144160fc40b6af919df306e82ea35e26ae81441e1a873f6e031113266dacbd7c  guix-build-568fcdddaec2/output/x86_64-w64-mingw32/bitcoin-568fcdddaec2-win64-setup-unsigned.exe
f1bfcca2a1bed1ceaa917c53a6ec41b215d00883839fa4b1f93f9f172b160a92  guix-build-568fcdddaec2/output/x86_64-w64-mingw32/bitcoin-568fcdddaec2-win64-unsigned.tar.gz
01ee17316684cecb30b2ec73d958abfa08cddc84941799116ee51276ae4afa55  guix-build-568fcdddaec2/output/x86_64-w64-mingw32/bitcoin-568fcdddaec2-win64.zip
```

ryanofsky reviewed (approved):
Code review ACK 568fcdddaec2cc8decba5a098257f31729cc1caa. Only change since last review was rebasing. I'm ok with this PR in its current form if other developers are happy with it. I just personally think it is inappropriate to \*silently\* break an everyday developer workflow like `git pull; make bitcoind`. I wouldn't have a problem with this PR if it triggered an explicit error, or if the problem was limited to less common workflows like changing cmake options in an existing build. 

I just think the silent failures this PR causes are unnecessary given (1) how easy it it to provide compatibility with a shim that doesn't touch the rest of the build, (2) how the shim provides useful functionality on its own and (3) how the shim can be safely removed later since removing it would not trigger silent failures.

This is just my opinion though. If it is not convincing I am ok with current approach.

sedited reviewed (approved):
Re-ACK 568fcdddaec2cc8decba5a098257f31729cc1caa

Sjors commented:
If I understand the remaining criticism correctly, it seems best to merge this before branch-off and improve things later.

The number of users that will be confused if we delay this change until _after_ branch-off is potentially much larger than the number of "everyday developer" folks impacted by the silent breakage here. For people who build the release from source, it's better if we educate them about the cmake changes and the new binary locations at the same time.

> the silent failures this PR causes are unnecessary given (1) how easy it it to provide compatibility with a shim that doesn't touch the rest of the build

To not lose the existing review work here, maybe someone can PR this right after branch-off?

theuni reviewed (approved):
ACK 568fcdddaec2cc8decba5a098257f31729cc1caa

+1 sjors.

I understand the arguments for/against the symlink, but don't feel strongly either way myself.

ryanofsky commented:
I don't see a good reason to cause silent breakage here, but if that's what we want to do it seems fine (really) to merge 568fcdddaec2cc8decba5a098257f31729cc1caa anytime. Would just point to https://github.com/bitcoin/bitcoin/pull/31161#pullrequestreview-2644640989 to explain why silent failures are unnecessary and should be easy to avoid.

l0rinc commented on doc/build-osx.md:
```diff
@@ -230,8 +230,8 @@ tail -f $HOME/Library/Application\ Support/Bitcoin/debug.log
+./build/bin/bitcoind -daemon      # Starts the bitcoin daemon.
+./build/bin/bitcoin-cli --help    # Outputs a list of command-line options.
+./build/bin/bitcoin-cli help      # Outputs a list of RPC commands when the daemon is running.
+./build/bin/bitcoin-qt -server # Starts the bitcoin-qt server mode, allows bitcoin-cli control
```
post-merge nit: if you touch again please realign the comments

l0rinc reviewed (commented):
post-merge ACK

Sjors commented:
This needed a release note :-)

Maybe we can still mention it in the next v29 update?

fanquake commented:
https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-29.0.md#build-system

> By default, the built executables and libraries are located in the bin/ and lib/ subdirectories of the build directory.
